### PR TITLE
Add DeepVoice3

### PR DESCRIPTION
**Pull request by kundanapillari**
# DeepVoice3

## Description
Deep learning model that performs end-to-end neural speech synthesis. Requires fewer parameters and is faster than other state-of-the-art neural speech synthesis systems.

## Model

|Model        |Download  | Download (with sample test data)|ONNX version|Opset version|Accuracy|
|-------------|:--------------|:--------------|:--------------|:--------------|:--------------|
|DeepVoice3      | [52.8 MB](audio/deepvoice3/model/deepvoice3-11.onnx) | [49.1 MB](audio/deepvoice3/model/deepvoice3-11.tar.gz) |1.6| 11|3.63 MOS |

### Source
PyTorch DeepVoice3 => ONNX DeepVoice3

## Inference
Used [ONNX Runtime](https://github.com/microsoft/onnxruntime) to perform inference

### Input
Sequence of words as a string in a list. Example: texts = ["Thank you very much.", "Hello.", "Deep voice 3."]. Inputs texts are converted into features including sequence resized to the shape [3,21] of type int64 and text_positions to the shape [3,12,80] of type float32. 

### Preprocessing
A few actions were taken on the text to improve the quality of the audio prediction:
* All Uppercase letters.
* Punctuation marks removed.
* Every utterance ended with a period or a question mark.
* Spaces replaced between words with separator characters that represent the time between words uttered by the speaker.

Get parameters and convert textual features to an internal representation:
``` 
sequence = np.array(frontend.text_to_sequence(text, p=p))
sequence = torch.from_numpy(sequence).unsqueeze(0).long().to(device)
text_positions = torch.arange(1, sequence.size(-1) + 1).unsqueeze(0).long().to(device)
speaker_ids = None if speaker_id is None else torch.LongTensor([speaker_id]).to(device)
```

### Output
Generates audio output. The decoder uses mel-band log-magnitude spectograms as audio frame representation.

### Postprocessing
Print text with generated audio:
```
for idx, text in enumerate(texts):
  print(idx, text)
  tts(model, text, figures=False)
```

### Dataset (Train and validation)
The original model is pretrained with the [LJSpeech](https://keithito.com/LJ-Speech-Dataset/) dataset. This is a public domain speech dataset consisting of 13,100 short audio clips of a single speaker reading passages from 7 non-fiction books.

### Training
The pretrained model is referenced in the [deepvoice3_pytorch](https://github.com/r9y9/deepvoice3_pytorch) repository. 

### Validation accuracy
Mean Opinion Score (MOS) of 3.78 computed over LJSpeech dataset.
<hr>

### References
Deep Voice3 model from the paper [Deep Voice 3: Scaling Text-to-Speech with Convolutional Sequence Learning](https://arxiv.org/abs/1710.07654)

## Contributors
Kundana Pillari 

## License
MIT "Expat" License 
<hr>

### CODE REVIEW

File: audio/deepvoice3/dependencies/text/symbols.py [hunk 1]
```diff
@@ -0,0 +1,17 @@
+'''
+Defines the set of symbols used in text input to the model.
+
+The default is a set of ASCII characters that works well for English or text that has been run
+through Unidecode. For other data, you can modify _characters. See TRAINING_DATA.md for details.
```
Comment (vinitra) [inline audio/deepvoice3/dependencies/text/symbols.py:5]: Remove references to TRAINING_DATA.md.
Also, if these files are borrowed from the source repository, please include a copyright disclaimer in the beginning.
i.e. "This file was originally created by the developers of the DeepVoice3 repository and is used to preprocess input data."

File: audio/deepvoice3/dependencies/frontend.py [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+    try:
+        phonemes = _arpabet[word][0]
+        phonemes = " ".join(phonemes)
+    except KeyError:
+        return word
+
+    return '{%s}' % phonemes if random() < p else word
+
+
+def mix_pronunciation(text, p):
```
Comment (vinitra) [inline audio/deepvoice3/dependencies/frontend.py:22]: Add a comment describing what this function does.

File: audio/deepvoice3/dependencies/frontend.py [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+    text = ' '.join(_maybe_get_arpabet(word, p) for word in text.split(' '))
+    return text
+
+
+def text_to_sequence(text, p=0.0):
```
Comment (vinitra) [inline audio/deepvoice3/dependencies/frontend.py:27]: Add a comment describing what this function does and what the arguments mean.

File: audio/deepvoice3/dependencies/text/cleaners.py [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+    ('hon', 'honorable'),
+    ('sgt', 'sergeant'),
+    ('capt', 'captain'),
+    ('esq', 'esquire'),
+    ('ltd', 'limited'),
+    ('col', 'colonel'),
+    ('ft', 'fort'),
+]]
+
+
```
Comment (vinitra) [inline audio/deepvoice3/dependencies/text/cleaners.py:43]: add comments with a quick description of the functions in this code (along the lines of the comment for basic_cleaners function below)

File: audio/deepvoice3/dependencies/text/numbers.py [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+            return 'two thousand ' + _inflect.number_to_words(num % 100)
+        elif num % 100 == 0:
+            return _inflect.number_to_words(num // 100) + ' hundred'
+        else:
+            return _inflect.number_to_words(num, andword='', zero='oh', group=2).replace(', ', ' ')
+    else:
+        return _inflect.number_to_words(num, andword='')
+
+
+def normalize_numbers(text):
```
Comment (vinitra) [inline audio/deepvoice3/dependencies/text/numbers.py:64]: function description

File: README.md [hunk 6]
```diff
@@ -99,7 +99,7 @@ This class of models uses audio data to train models that can identify voice, ge
 |Model Class |Reference |Description |
 |-|-|-|
 |Speech recognition with deep recurrent neural networks|	[Graves et al.](https://www.cs.toronto.edu/~fritz/absps/RNN13.pdf)|A RNN model for sequential data for speech recognition. Labels problems where the input-output alignment is unknown<br>[contribute](contribute.md)|
-|Deep voice: Real time neural text to speech |	[Arik et al.](https://arxiv.org/abs/1702.07825)	|A DNN model that performs end-to-end neural speech synthesis. Requires fewer parameters and it is faster than other systems. <br>[contribute](contribute.md)|
+|<b>[Deep voice: Real time neural text to speech](audio/deepvoice3)</b> |	[Arik et al.](https://arxiv.org/abs/1702.07825)	|A DNN model that performs end-to-end neural speech synthesis. Requires fewer parameters and it is faster than other systems. <br>[contribute](contribute.md)|
```
Comment (vinitra) [inline README.md:102]: nit: A deep learning model that
nit: Requires fewer parameters and is faster
describe which systems it's faster than

Comment (vinitra) [reply README.md:102]: Also, remove "contribute.md" file.

File: audio/deepvoice3/dependencies/text/symbols.py (hunk 1 above)
Comment (kundanapillari) [reply audio/deepvoice3/dependencies/text/symbols.py]: Made the change and also removed the pycache files.

File: audio/deepvoice3/README.md [hunk 7]
```diff
@@ -0,0 +1,67 @@
+# DeepVoice3
+
+## Description
+Deep learning model that performs end-to-end neural speech synthesis. Requires fewer parameters and is faster than other state-of-the-art neural speech synthesis systems.
+
+## Model
+
+|Model        |Download  | Download (with sample test data)|ONNX version|Opset version|Accuracy|
+|-------------|:--------------|:--------------|:--------------|:--------------|:--------------|
+|DeepVoice3      | [52.8 MB](audio/deepvoice3/model/deepvoice3-11.onnx) | [49.1 MB](audio/deepvoice3/model/deepvoice3-11.tar.gz) |1.6| 11|Speech naturalness of 95.5 Pronunciation accuracy of 95.1 |
```
Comment (vinitra) [inline audio/deepvoice3/README.md:10]: Accuracy portion renders weirdly. Are there actual metric names mentioned in the paper for "speech naturalness" and Pronunciation accuracy?

```suggestion
|DeepVoice3      | [52.8 MB](audio/deepvoice3/model/deepvoice3-11.onnx) | [49.1 MB](audio/deepvoice3/model/deepvoice3-11.tar.gz) |1.6| 11|Speech=95.5, Pronunciation=95.1 |
```

File: audio/deepvoice3/README.md [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+
+### Source
+Pytorch => onnx model
```
Comment (vinitra) [inline audio/deepvoice3/README.md:13]:
```suggestion
PyTorch DeepVoice3 => ONNX DeepVoice3
```

File: audio/deepvoice3/README.md [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+
+## Inference
+Used [ONNX Runtime](https://github.com/microsoft/onnxruntime) to perform inference
```
Comment (vinitra) [inline audio/deepvoice3/README.md:16]: It would be nice if you could include an inference sample here, like in Jennifer's PR: https://github.com/onnx/models/pull/322

File: audio/deepvoice3/README.md [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+Sequence of words as a string in a list. Example: texts = ["Thank you very much.", "Hello.", "Deep voice 3."]. Inputs texts are converted into features including sequence resized to the shape [3,21] of type int64 and text_positions to the shape [3,12,80] of type float32. 
+
+### Preprocessing
+A few actions were taken on the text to improve the quality of the audio prediction:
+* All Uppercase letters.
+* Punctuation marks removed.
+* Every utterance ended with a period or a question mark.
+* Spaces replaced between words with separator characters that represent the time between words uttered by the speaker.
+
+Get parameters and convert textual features to an internal representation.
```
Comment (vinitra) [inline audio/deepvoice3/README.md:28]: nit: use colon at end of line instead of period

File: audio/deepvoice3/README.md (hunk 7 above)
Comment (kundanapillari) [reply audio/deepvoice3/README.md]: I looked more into the paper and changed the scoring to Mean Opinion Score (MOS) of 3.78

**Comment by kundanapillari** [comment]
The Decoder method of the model causes some issues when converting the model into ONNX format. 
More information can be found here: https://github.com/pytorch/pytorch/issues/41304
